cranelift/codegen/src/isa/x64/lower/isle/generated_code.rs

```diff
@@ -0,0 +1,2608 @@
+            src1: expr1_0,
+            src2: expr4_0,
+            dst: expr0_0,
+        };
+        let expr6_0 = C::emit(ctx, &expr5_0);
+        return Some(expr1_0);
+    }
+    // Rule at src/isa/x64/inst.isle line 565.
+    let expr0_0 = C::temp_writable_reg(ctx, pattern0_0);
+    let expr1_0 = C::operand_size_of_type(ctx, pattern0_0);
+    let expr2_0 = MInst::Imm {
+        dst_size: expr1_0,
+        simm64: pattern1_0,
+        dst: expr0_0,
+    };
+    let expr3_0 = C::emit(ctx, &expr2_0);
+    let expr4_0 = C::writable_reg_to_reg(ctx, expr0_0);
+    return Some(expr4_0);
+}
+
+// Generated as internal constructor for term shift_r.
+pub fn constructor_shift_r<C: Context>(ctx: &mut C, arg0: Type, arg1: &ShiftKind, arg2: Reg, arg3: &Imm8Reg) -> Option<Reg> {
+    let pattern0_0 = arg0;
+    let pattern1_0 = arg1;
+    let pattern2_0 = arg2;
+    let pattern3_0 = arg3;
+    // Rule at src/isa/x64/inst.isle line 595.
+    let expr0_0 = C::temp_writable_reg(ctx, pattern0_0);
+    let expr1_0 = C::operand_size_of_type(ctx, pattern0_0);
+    let expr2_0 = MInst::ShiftR {
+        size: expr1_0,
+        kind: pattern1_0.clone(),
+        src: pattern2_0,
+        num_bits: pattern3_0.clone(),
+        dst: expr0_0,
+    };
+    let expr3_0 = C::emit(ctx, &expr2_0);
+    let expr4_0 = C::writable_reg_to_reg(ctx, expr0_0);
+    return Some(expr4_0);
+}
+
+// Generated as internal constructor for term m_rotl.
+pub fn constructor_m_rotl<C: Context>(ctx: &mut C, arg0: Type, arg1: Reg, arg2: &Imm8Reg) -> Option<Reg> {
+    let pattern0_0 = arg0;
+    let pattern1_0 = arg1;
+    let pattern2_0 = arg2;
+    // Rule at src/isa/x64/inst.isle line 604.
+    let expr0_0 = ShiftKind::RotateLeft;
+    let expr1_0 = constructor_shift_r(ctx, pattern0_0, &expr0_0, pattern1_0, pattern2_0)?;
+    return Some(expr1_0);
+}
+
+// Generated as internal constructor for term shl.
+pub fn constructor_shl<C: Context>(ctx: &mut C, arg0: Type, arg1: Reg, arg2: &Imm8Reg) -> Option<Reg> {
+    let pattern0_0 = arg0;
+    let pattern1_0 = arg1;
+    let pattern2_0 = arg2;
+    // Rule at src/isa/x64/inst.isle line 609.
+    let expr0_0 = ShiftKind::ShiftLeft;
+    let expr1_0 = constructor_shift_r(ctx, pattern0_0, &expr0_0, pattern1_0, pattern2_0)?;
+    return Some(expr1_0);
+}
+
+// Generated as internal constructor for term shr.
+pub fn constructor_shr<C: Context>(ctx: &mut C, arg0: Type, arg1: Reg, arg2: &Imm8Reg) -> Option<Reg> {
+    let pattern0_0 = arg0;
+    let pattern1_0 = arg1;
+    let pattern2_0 = arg2;
+    // Rule at src/isa/x64/inst.isle line 614.
+    let expr0_0 = ShiftKind::ShiftRightLogical;
+    let expr1_0 = constructor_shift_r(ctx, pattern0_0, &expr0_0, pattern1_0, pattern2_0)?;
+    return Some(expr1_0);
+}
+
+// Generated as internal constructor for term sar.
+pub fn constructor_sar<C: Context>(ctx: &mut C, arg0: Type, arg1: Reg, arg2: &Imm8Reg) -> Option<Reg> {
+    let pattern0_0 = arg0;
+    let pattern1_0 = arg1;
+    let pattern2_0 = arg2;
```

Comment:
It's true that the DSL compiler has a simple emission strategy, but I don't think this sort of thing specifically actually results in worse code. Basically we are taking a compound expression and pulling it apart into separate let-bindings; afaik the MIR from this should be identical.

Reply:
MIR doesn't do copy propagation. Every assignment results in a copy at MIR level.

Current code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8cf97ba023221ffafccb62ab89231fab

Excerpt:

```rust
    bb0: {
        _5 = move _2;                    // scope 0 at src/lib.rs:21:22: 21:26
        _6 = move _3;                    // scope 1 at src/lib.rs:22:22: 22:26
        _7 = _4;                         // scope 2 at src/lib.rs:23:22: 23:26
        discriminant(_8) = 0;            // scope 3 at src/lib.rs:25:19: 25:50
        _12 = &mut (*_1);                // scope 4 at src/lib.rs:26:39: 26:42
        _13 = move _5;                   // scope 4 at src/lib.rs:26:44: 26:54
        _15 = &_8;                       // scope 4 at src/lib.rs:26:56: 26:64
        _14 = _15;                       // scope 4 at src/lib.rs:26:56: 26:64
        _16 = move _6;                   // scope 4 at src/lib.rs:26:66: 26:76
        _17 = _7;                        // scope 4 at src/lib.rs:26:78: 26:88
        _11 = constructor_shift_r::<C>(move _12, move _13, move _14, move _16, move _17) -> bb1; // scope 4 at src/lib.rs:26:19: 26:89
                                         // mir::Constant
                                         // + span: src/lib.rs:26:19: 26:38
                                         // + literal: Const { ty: for<'r, 's, 't0> fn(&'r mut C, Type, &'s ShiftKind, Reg, &'t0 Imm8Reg) -> std::option::Option<Reg> {constructor_shift_r::<C>}, val: Value(Scalar(<ZST>)) }
    }
```

With copies eliminated: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=61312e466966b68bd721e61f98cc62e3

```rust
    bb0: {
        _8 = &mut (*_1);                 // scope 0 at src/lib.rs:21:39: 21:42
        _9 = move _2;                    // scope 0 at src/lib.rs:21:44: 21:48
        _19 = const constructor_sar::<C>::promoted[0]; // scope 0 at src/lib.rs:21:50: 21:82
        _11 = _19;                       // scope 0 at src/lib.rs:21:50: 21:82
        _10 = _11;                       // scope 0 at src/lib.rs:21:50: 21:82
        _12 = move _3;                   // scope 0 at src/lib.rs:21:84: 21:88
        _13 = _4;                        // scope 0 at src/lib.rs:21:90: 21:94
        _7 = constructor_shift_r::<C>(move _8, move _9, move _10, move _12, move _13) -> bb1; // scope 0 at src/lib.rs:21:19: 21:95
                                         // mir::Constant
                                         // + span: src/lib.rs:21:19: 21:38
                                         // + literal: Const { ty: for<'r, 's, 't0> fn(&'r mut C, Type, &'s ShiftKind, Reg, &'t0 Imm8Reg) -> std::option::Option<Reg> {constructor_shift_r::<C>}, val: Value(Scalar(<ZST>)) }
    }
```

Something that would probably help a lot more though is avoiding `?` and instead using `match`. `?` desugars into several function calls:

```rust
    bb1: {
        _7 = <Option<Reg> as Try>::branch(move _8) -> bb2; // scope 1 at src/lib.rs:22:19: 22:68
                                         // mir::Constant
                                         // + span: src/lib.rs:22:19: 22:68
                                         // + literal: Const { ty: fn(std::option::Option<Reg>) -> std::ops::ControlFlow<<std::option::Option<Reg> as std::ops::Try>::Residual, <std::option::Option<Reg> as std::ops::Try>::Output> {<std::option::Option<Reg> as std::ops::Try>::branch}, val: Value(Scalar(<ZST>)) }
    }

    bb2: {
        _15 = discriminant(_7);          // scope 1 at src/lib.rs:22:19: 22:68
        switchInt(move _15) -> [0_isize: bb3, 1_isize: bb5, otherwise: bb4]; // scope 1 at src/lib.rs:22:19: 22:68
    }

    bb3: {
        _18 = move ((_7 as Continue).0: Reg); // scope 1 at src/lib.rs:22:19: 22:68
        _6 = move _18;                   // scope 6 at src/lib.rs:22:19: 22:68
        _19 = move _6;                   // scope 2 at src/lib.rs:23:17: 23:24
        ((_0 as Some).0: Reg) = move _19; // scope 2 at src/lib.rs:23:12: 23:25
        discriminant(_0) = 1;            // scope 2 at src/lib.rs:23:12: 23:25
        goto -> bb6;                     // scope 2 at src/lib.rs:23:5: 23:25
    }

    bb4: {
        unreachable;                     // scope 1 at src/lib.rs:22:19: 22:68
    }

    bb5: {
        _0 = <Option<Reg> as FromResidual<Option<Infallible>>>::from_residual(move _17) -> bb6; // scope 4 at src/lib.rs:22:19: 22:68
                                         // mir::Constant
                                         // + span: src/lib.rs:22:67: 22:68
                                         // + literal: Const { ty: fn(std::option::Option<std::convert::Infallible>) -> std::option::Option<Reg> {<std::option::Option<Reg> as std::ops::FromResidual<std::option::Option<std::convert::Infallible>>>::from_residual}, val: Value(Scalar(<ZST>)) }
    }

    bb6: {
        return;                          // scope 0 at src/lib.rs:24:2: 24:2
    }
```

Reply:
OK, interesting. Echoing the point I made above in another thread, though, this seems to me to be more a problem that debug builds of Rust do no optimization whatsoever (and this is a problem for *any* Rust code that is written with abstractions or redundancies that optimize away), and less that our particular DSL codegen produces code that has optimizable parts.

Reply:
I agree that this kind of thing could end up being significant, especially as we add more patterns to isle and the generated files grow. However, optimizations like these complicate the generator, and right now, keeping that code simple is particularly valuable, as the code is new and still rapidly changing. 

As ISLE matures, we add more patterns, and the generated files become large enough that effects like this are measurable, I expect it'll make sense to revisit this.